### ISSUE TITLE
Add regex support for id replacement?

### ISSUE THREAD

**Issue by cissoid** [naokazuterada/MarkdownTOC#56]
This plugin is so great, but I have some problem to use this plugin with GitLab.

GitLab has its own markdown syntax called [GitLab Flavored Markdown](http://doc.gitlab.com/ce/markdown/markdown.html). When render markdown into html, there are several problem:
1. Header's anchor id is generated by its value, but all non-ascii characters(like chinese character) will be trimmed. For example:
   `# 1. Hello World` will has an id `1-hello-world`
   `# 1. Hello 世界 World` will also has id `1-hello-world`, "世界" is trimmed.
   So if I want this plugin support gitlab, I must add all chinese characters(6,000 at least) to `id_replacements` settings, it's impossible.
2. Another way is set autoanchor=true, this will add an `<a name="anchor-id"></a>`, but I don't know why, this tag is ignored when render to html.

So I thought that maybe `id_replacements` could add regex support, so that I could use just one regex to match all characters I need, just like this:

``` python
for _substitute, pattern in regex_replacements.iteritems():
    _str = re.sub(pattern, _substitute, _str)
```

**Comment by lahmatiy** [comment]
+1
But regexp doesn't needed. Just non-ascii characters should be escaped, like github/gitlab does. It's may be an optional.
In JavaScript escaping may be done by using `encodeURIComponent(str)`. In Python it's also simple (according to [stackoverflow](http://stackoverflow.com/questions/946170/equivalent-javascript-functions-for-pythons-urllib-quote-and-urllib-unquote)) – `urllib.quote(unicode(str).encode('utf-8'), safe='~()*!.\'')`

**Comment by lahmatiy** [comment]
I tried to use plugin again – it doesn't work to me still. Since it lowercases non-latin chars and links in TOC don't work if they contains non-latin :(

**Comment by naokazuterada** [comment]
@cissoid

I'm sooooooo sorry for late reply. 🤕 

Today I added a new feature [URI encoding](https://github.com/naokazuterada/MarkdownTOC#uri-encoding). And I checked links working on Gitlab and Github. Here's samples [Gitlab](https://gitlab.com/naokazu.terada/MarkdownTOC-Tests/blob/f8f1f479589dcb02f9e1262e77a41b0d49f1985e/README.md), [Github](https://github.com/naokazuterada/MarkdownTOC-Tests/blob/b8f1c826b154d9d2a25608f081b31156cb5c2abd/README.md). I hope this fix your problem. 🦊  I'm looking forward to your response if you don't mind.

> `# 1. Hello 世界 World` will also has id `1-hello-world`, "世界" is trimmed.

By the way, I think it won't be `1-hello-world` but `1-hello--world` because `世界` is trimmed. So I think you will not be able to get results you wanted with 'id_replacements' feature anyway.

**Comment by naokazuterada** [comment]
This may be fixed in [2.4.0](https://github.com/naokazuterada/MarkdownTOC/releases/tag/2.4.0).
